include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -573,6 +575,45 @@ class __future : private std::tuple<_Args...>
     }
 };
 
+// A contract for a future class for reduce: <execution policy, sycl::event, USM host memory for the reduced value>
+template <typename _ExecutionPolicy, typename _Event, typename _Res>
+class __reduce_future
```

Comment:
I think we don't need the same addition feature type.
We may generalize existing one  - class __future, by  passing instead of buffer some RAII type based on ::std::unique_ptr. f.e.
```
template<typename Res>
using usm_value_buf = ::std::unique_ptr<_Res, ::std::function<void(_Res*)>>;
auto make_usm_value_buf(queue q, Res* res) { return usm_value_buf(res, [q](_Res* __res) { ::sycl::free(res, queue); }); }
```

Reply:
I don't suggest to have type-erased utility (like `std::function`). Let's avoid allocation we don't control. You can always write deleter as a named class (not as a lambda) so I don't see why this overhead is necessary.

In general I am in favor of genericity idea. We always use either buffer of pointer, right? Can we just have compile-time check and rely that pointer always was allocated via USM?

Reply:
Usage of ```std::function``` were removed from this PR.